### ISSUE TITLE
Documentation: add SQL support via objection-authorize

### ISSUE THREAD

**Issue by JaneJeon** [stalniy/casl#486]
Hi! I am a daily user of `casl` and over the past few years I've been using `casl` along with `objection-authorize` (a library that I've also been building for the past few years) to integrate `casl` in both the frontend and the backend isomorphically, and perhaps more importantly, ACTUALLY use the `casl` ACLs to filter/check SQL queries!

I understand that #8 is asking for *Sequelize* support, but it doesn't lend itself to plugin support very neatly, and I'm a personal advocate of Objection.js (the SQL ORM of choice), and perhaps most importantly, it's ALREADY proven to be working (see https://github.com/JaneJeon/blink for an example, especially under `policies/` and `models/` directory).

https://github.com/JaneJeon/objection-authorize has a full battery of tests that have been expanding over the years, and at this point I feel that it's "battle-tested" enough to recommend to the general public. So I'd appreciate it if you could add `objection-authorize` as an adapter for people using SQL + CASL, under the "Official Packages" in the documentation.

My hope is that with my library, people who're looking for SQL support *don't* have to overlook CASL because of the lack of support for SQL ORMs (and vice versa).

Thanks!

**Comment by stalniy** [comment]
Hi @JaneJeon 

Thanks for letting me know! This is awesome! However, I need to take a closer look to your library before I can recommend it on casl resources.

**Comment by JaneJeon** [comment]
Yeah, feel free to read through the README and the tests, and if you're still not convinced, please do look into how blink uses it to achieve isomorphism!

**Comment by JaneJeon** [comment]
@stalniy did you get to check it out yet?

**Comment by stalniy** [comment]
Sorry @JaneJeon when you have a little kid Open Source activities start to move a bit slower :). I definitely will take a look but it will take some time

**Comment by JaneJeon** [comment]
@stalniy any updates?

**Comment by JaneJeon** [comment]
@stalniy that’s fine, I just thought I’d poke you now since it seemed like you were active again on this repo recently

**Comment by stalniy** [comment]
OK, I took a look. And have few questions:

1. As far as I can understand the code, permissions are checked in model hooks, right?
2. How do you fetch accessible records from database? For example, I want to get all articles that I can read from the database

**Comment by JaneJeon** [comment]
As for the particular query you asked about (https://github.com/JaneJeon/blink/blob/master/routes/api/users.js#L8), let’s first disregard the authorization and see what the query is trying to do.

The user is trying to read a certain set (max 15, this is a paginated query) of links. You *could* also just call `.fetchResourceContextFromDB()`, but either way, you’re going to have to make the call to the DB in order to check “is User X allowed to read Link Y”, so in GET requests like those, you have an additional option of taking the result of the query and checking it synchronously using `.authorizeRead(req.user)` (which will also automatically strip out any fields that you’re not authorized to read according to the policies).

**Comment by JaneJeon** [comment]
TL;DR: it is somewhat similar to casl-mongoose/prisma, and yet different in its purpose to drive isomorphic authorization logic and use existing query context to make authorization decisions, so it does more than just checking its given inputs against a given policy - it already KNOWS which policy and which action you’re trying to take (however, note that it is not “magic”, as you can specify literally every part of the authorization decision within the query - the action, the user, the resource, and additional context).

**Comment by stalniy** [comment]
OK, just to double check, if I have such policy

```js
allow('read', 'User', { id: user.id })
```

And then want to authorize a list of users:

```js
const users = await User.query().authorize(req.user)
```

What will I happen if there are 1000 of users in db?

**Comment by JaneJeon** [comment]
Ah, so you’re worried about the potential impact of the DB. No worries there.

1. The query that you wrote would literally fetch all of the users in the DB. If you want that, sure. If you want to check the authorization gainst *individual* user, you could do `.authorizeRead()` on the resulting set.
2. The plugin doesn’t “by default” fetch all of the resources - it is an EXPLICIT call using `.fetchResourceContextFromDB()` - so in this case, it would just check if the `req.user` had the access to `read` a `User` in general, *without* making additional DB calls

**Comment by JaneJeon** [comment]
`.fetchResourceContextFromDB()` is really meant for cases where you’re doing a PUT request, for example, and you need to check what the user *says* is the modified model instance versus what is ACTUALLY in the database. In cases like that, you would need to actually fetch the record from the database.

**Comment by stalniy** [comment]
OK, I see. Thanks for the explanation!

After some thinking, I realized that there may be other plugins for CASL, I'm and others are not aware of. So, I think a better strategy would be to ask package authors to add `casl` keyword inside package.json. So, then it's easily searchable on npm -> https://www.npmjs.com/search?q=keywords:casl . This is how mongoose does this and it makes sense for me as well.

I'll create a page in docs that explains that.

**Comment by JaneJeon** [comment]
@stalniy also, it would be especially helpful if you had different sections for 3rd party plugins, basically grouping them generally in terms of what they do (e.g. “frontend plugins”, “backend plugins”, “CASL query transformers”, etc).

**Comment by stalniy** [comment]
@JaneJeon I added a note about 3rd part packages.

I'm sorry but I don't want to maintain that list. Once somebody adds something to my code, I will need to support it. When amount of that packages grow and everybody will want to put own package in docs, it will become a nightmare for me because:

1. Review that PRs
2. Ensure that packages are alive and remove if some of them become abandoned 

I think that I can spend my time more effectively :)

**Comment by JaneJeon** [comment]
Well, I’m not sure that maintaining 6 packages on your own when you don’t have the time to respond to a simple issue is a “more effective” use of your time (esp. since Objection.js just pushes that PR reviewing/ensuring packages are alive part to the community, because *users will let you know* if a package is abandoned), but if you want to go that route, I wish you the best of luck.

**Comment by stalniy** [comment]
Ouch, not nice :)
